File: server/src/models/Room.ts
import mongoose, { Document, Schema, Types } from 'mongoose';
import { generateRoomId } from '../utils/roomIdGenerator';

// TypeScript interface for Room
// User in room metadata
export interface IRoomUser {
  userId: Types.ObjectId;
  userName: string;
  role: 'owner' | 'collaborator';
  joinedAt: Date;
}

export interface IRoom extends Document {
  roomId: string;
  name: string;
  ownerId: Types.ObjectId;
  language: 'javascript' | 'python' | 'java' | 'cpp' | 'typescript' | 'go' | 'rust';
  code: string;
  collaborators: Types.ObjectId[];
  users: IRoomUser[]; // Active users with metadata
  maxUsers: number; // Maximum allowed users
  isPublic: boolean;
  isActive: boolean; // Room active status
  createdAt: Date;
  updatedAt: Date;
}

// Schema definition
const roomSchema = new Schema<IRoom>(
  {
    roomId: {
      type: String,
      required: true,
      unique: true,
    },
    name: {
      type: String,
      required: [true, 'Room name is required'],
      trim: true,
    },
    ownerId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Owner ID is required'],
    },
    language: {
      type: String,
      enum: ['javascript', 'python', 'java', 'cpp', 'typescript', 'go', 'rust'],
      default: 'javascript',
    },
    code: {
      type: String,
      default: '// Start coding...',
    },
    collaborators: [
      {
        type: Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    users: [
      {
        userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
        userName: { type: String, required: true },
        role: { type: String, enum: ['owner', 'collaborator'], required: true },
        joinedAt: { type: Date, default: Date.now },
      },
    ],
    maxUsers: {
      type: Number,
      default: 50,
    },
    isPublic: {
      type: Boolean,
      default: false,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
  },
  { timestamps: true }
);

// ✅ Generate unique roomId if not already set
roomSchema.pre('validate', async function (next) {
  if (!this.roomId) {
    const RoomModel = mongoose.model<IRoom>('Room');
    let id: string;
    let unique = false;

    while (!unique) {
      id = generateRoomId(); // ABC-123 format
      const existing = await RoomModel.findOne({ roomId: id });
      if (!existing) unique = true;
    }

    this.roomId = id!;
  }
  next();
});

// Index for fast lookup
roomSchema.index({ roomId: 1 }, { unique: true });

export const Room = mongoose.model<IRoom>('Room', roomSchema);

File: server-spring/src/main/java/com/codeeditor/server/model/Room.java
package com.codeeditor.server.model;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.FieldType;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "rooms")
public class Room {

    @Id
    private String id;

    @Indexed(unique = true)
    private String roomId;

    private String name;

    @Field(targetType = FieldType.OBJECT_ID)
    private String ownerId; // References User._id

    private String language = "javascript"; // javascript, python, java, cpp, typescript, go, rust

    private String code = "// Start coding...";

    @Field(targetType = FieldType.OBJECT_ID)
    private List<String> collaborators = new ArrayList<>(); // List of User._id

    private List<RoomUser> users = new ArrayList<>();

    private int maxUsers = 50;

    private boolean isPublic = false;

    private boolean isActive = true;

    private boolean requireApproval = true;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class RoomUser {
        @Field(targetType = FieldType.OBJECT_ID)
        private String userId;
        private String userName;
        private String role; // "owner" or "collaborator"
        private Instant joinedAt = Instant.now();
    }
}
